src/graph/impl/KokkosGraph_Distance2Color_impl.hpp

```diff
@@ -41,898 +41,1453 @@
-        for(size_type vid_2adj=h_t_idx(vid_1idx); vid_2adj < h_t_idx(vid_1idx+1); vid_2adj++)
+        swap_work_arrays              = true;
+        nnz_lno_t output_numUncolored = 0;
+
+        // conflictList mode:
+        if(0 == this->_conflictList)
         {
-          nnz_lno_t vid_2idx = h_t_adj(vid_2adj);
+            // Throw an error -- not implemented (yet)
+            std::ostringstream os;
+            os << "GraphColorD2::findConflicts() not implemented for conflictList == 0";
+            Kokkos::Impl::throw_runtime_exception(os.str());
+        }
 
-          // skip over loops vid -- x -- vid
-          if(vid_2idx == vid)
-            continue;
+        // conflictList mode: Parallel Prefix Sums (PPS)
+        else if(2 == this->_conflictList)
+        {
+            // Throw an error -- not implemented (yet)
+            std::ostringstream os;
+            os << "GraphColorD2::findConflicts() not implemented for conflictList == 2";
+            Kokkos::Impl::throw_runtime_exception(os.str());
+        }
 
-          forbidden[h_colors(vid_2idx)] = vid;
+        // conflictList mode: ATOMIC
+        else if(1 == this->_conflictList)
+        {
+            if(0 == this->_use_color_set)
+            {
+                functorFindConflicts_Atomic<adj_view_t> conf(this->nv, xadj_, adj_, t_xadj_, t_adj_, vertex_colors_, current_vertexList_, next_iteration_recolorList_, next_iteration_recolorListLength_);
+                Kokkos::parallel_reduce(my_exec_space(0, current_vertexListLength_), conf, output_numUncolored);
+            }
+        }
+        else
+        {
+            // Throw an error becaue we should not be here...
+            std::ostringstream os;
+            os << "GraphColorD2::findConflicts() - unknown conflictList Flag value: " << this->_conflictList << " ";
+            Kokkos::Impl::throw_runtime_exception(os.str());
         }
-      }
+        return output_numUncolored;
+    }      // findConflicts (end)
 
-      // color vertex vid with smallest available color
-      int c=1;
-      while (forbidden[c]==vid)  c++;
 
-      h_colors(vid) = c;
-    }
-    Kokkos::deep_copy(vertex_colors_, h_colors);
-    delete [] forbidden;
-  }  // resolveConflicts (end)
 
+    // -----------------------------------------------------------------
+    //
+    // GraphColorD2::resolveConflicts()
+    //
+    // -----------------------------------------------------------------
+    template<typename adj_view_t>
+    void resolveConflicts(nnz_lno_t _nv,
+                          const_lno_row_view_t xadj_,
+                          adj_view_t adj_,
+                          const_clno_row_view_t t_xadj_,
+                          const_clno_nnz_view_t t_adj_,
+                          color_view_type vertex_colors_,
+                          nnz_lno_temp_work_view_t current_vertexList_,
+                          size_type current_vertexListLength_)
+    {
+        color_t *forbidden = new color_t[_nv];
```

Comment:
My comment was about this line. _nv is #vertices, right ?

Reply:
Ah, ok, I see what you're looking at now... 

So in [resolveConflicts][2]:
```c++
    template<typename adj_view_t>
    void resolveConflicts(nnz_lno_t _nv,
                          const_lno_row_view_t xadj_,
                          adj_view_t adj_,
                          const_clno_row_view_t t_xadj_,
                          const_clno_nnz_view_t t_adj_,
                          color_view_type vertex_colors_,
                          nnz_lno_temp_work_view_t current_vertexList_,
                          size_type current_vertexListLength_)
    {
        color_t *forbidden = new color_t[_nv];
        nnz_lno_t vid      = 0;
        nnz_lno_t end      = _nv;
```

The function `resolveConflicts` is called after the iterative loop exits up on [line 310][1]:

```c++
// ------------------------------------------
// clean up in serial (resolveConflicts)
// ------------------------------------------
if(numUncolored > 0)
{
    this->resolveConflicts(this->nv, this->xadj, this->adj, this->t_xadj, this->t_adj, colors_out, current_vertexList, current_vertexListLength);
}
```
This is done in serial to clean up any issues that remain after the parallel coloring is done so in that case, the forbidden array does need to be O(V).  

From test runs on the Laplace_285 dataset that I sent you last week, `resolveConflicts` takes about 0.0008 seconds total across 10 runs on my laptop.  This was the "Total Time RC" column on the tables.

[1]: https://github.com/william76/kokkos-kernels/blob/328fede3b7bfc9ca97e132e912fdec736ad3f930/src/graph/impl/KokkosGraph_Distance2Color_impl.hpp#L305-L311
[2]: https://github.com/william76/kokkos-kernels/blob/328fede3b7bfc9ca97e132e912fdec736ad3f930/src/graph/impl/KokkosGraph_Distance2Color_impl.hpp#L538-L551